aten/src/ATen/native/TensorShape.cpp

```diff
@@ -28,6 +29,136 @@
 #include <vector>
 
 namespace at {
+namespace meta {
+static void cat_check_no_zero_dim(ITensorList tensors) {
+  for (const auto i : c10::irange(tensors.size())) {
+    auto& t = tensors[i];
+    TORCH_CHECK(
+        t.dim() > 0,
+        "zero-dimensional tensor (at position ", i, ") cannot be concatenated");
+  }
+}
+
+static c10::MemoryFormat cat_compute_output_memory_format(ITensorList inputs) {
+  c10::optional<c10::MemoryFormat> format = c10::nullopt;
+  for (auto& t : inputs) {
+    auto f = t.suggest_memory_format();
+    if (f == c10::MemoryFormat::Contiguous) {
+        return f;
+    }
+    if (format.has_value() && format.value() != f) {
+        return c10::MemoryFormat::Contiguous;
+    }
+    format = f;
+  }
+  return format.value();
+}
+
+TORCH_PRECOMPUTE_META_FUNC(cat)(ITensorList tensors, int64_t dim) {
+  // previously, size [0] tensors were the only possible empty tensors; thus, it wasn't possible
+  // to cat empty tensors unless all the other tensors were 1-dimensional, so we allowed these tensors
+  // to be "skipped".  We maintain this behavior for backwards compatibility, but only for this specific
+  // size (i.e. other empty sizes are not skipped).
+  cat_check_no_zero_dim(tensors);
+  dim = at::legacy_cat_wrap_dim(dim, tensors);
+
+  // Checking names before the actual dimensions.
+  auto maybe_outnames = namedinference::compute_cat_outnames(tensors);
+
+  TORCH_CHECK(
+      tensors.size() > 0, "torch.cat(): expected a non-empty list of Tensors");
+
+  // Look for the first valid tensor.
+  size_t valid = tensors.size();
+  for (const auto i : c10::irange(tensors.size())) {
+    if (!at::native::cat_should_skip_tensor(tensors[i])) {
+      valid = i;
+      break;
+    }
+  }
```

Comment:
Man the fact that `[0]` size tensors don't participate in size calculation is so janky. We should deprecate this and then make it just raise an error haha, then we could simplify this code quite a bit. (Nothing you need to do for this PR; if you agree maybe file an issue.)